### PR TITLE
Java bindings for Fixed-point type support for Parquet

### PR DESCRIPTION
**Pull request by razajafri**
Adds in java support to be able to write fixed-point type to parquet

### CODE REVIEW

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
I have removed the [skip ci] as this change has some very small CPP header file change

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:68]: Do we really want to support calling this method iteratively?  I could see users being confused, thinking this will override previous calls.

I think we either want this to override any previous call or take a single precision value and append it.  Personally I prefer the former.  That would also simplify the builder, as it can track the array directly.

Either way the Javadoc should be clear on the semantics of how the input will be handled on subsequent invocations.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:865]: Indentation is off here. 
```suggestion
    chunked_parquet_writer_options opts =
```

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:65]: This needs to elaborate on what "Flattened' means.  Concrete examples would help.  Also it should explain what happens if the number of specified decimal precisions doesn't cover all decimal columns being written or left completely unspecified.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:922]: Indentation is off here
```suggestion
    std::shared_ptr<pq_chunked_state> state = write_parquet_chunked_begin(opts);
```

File: cpp/include/cudf/io/parquet.hpp
Comment (jlowe) [inline cpp/include/cudf/io/parquet.hpp:846]: 2021 copyrights should be added.

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:18]: 2021 copyrights

File: java/src/main/java/ai/rapids/cudf/Table.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/Table.java:239]: 2021 copyrights

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:833]: 2021 copyrights

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:3949]: 2021 copyrights

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:3967]: This test and `testParquetWriteToBufferChunked` don't use the new decimals table.  Should those be updated to use this as well?

**Review by jlowe — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/io/parquet.hpp
Comment (revans2) [inline cpp/include/cudf/io/parquet.hpp:855]: How is this intended to work with nested types, like a list of decimal values? or a list of struct of decimal to long?

Also is a `uint8_t` really the maximum precision that parquet can support? I know it is above the maximum that we can support right now, and can handle 128-bit support too, but it would be good to know if parquet can support higher, jsut to future proof the API.

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:67]: I really hate this API as a user facing API. I really would have preferred to see us expose a parquet like schema API. In the future we are going to need to pass down other things like is this list a byte array or a binary value. Is this list of structs really a map. Can we try and spend a little time to put the beginnings of that in place that we could then extend later on?

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:96]: If this API is public we need to document exactly what it returns.

**Review by revans2 — COMMENTED** [review]
I am OK with letting the current decimal precision API through if we feel that there are time constraints, but I really do find it ugly and decidedly not forward thinking.

File: cpp/include/cudf/io/parquet.hpp
Comment (razajafri) [reply cpp/include/cudf/io/parquet.hpp:855]: > How is this intended to work with nested types, like a list of decimal values? or a list of struct of decimal to long?

I don't know if I am understanding your question correctly. Its a flattened list of precisions so a list of decimal types will only have one precision, @hyperbolic2346 can you correct me if I am wrong? 

> Also is a uint8_t really the maximum precision that parquet can support? I know it is above the maximum that we can support right now, and can handle 128-bit support too, but it would be good to know if parquet can support higher, jsut to future proof the API.

So parquet is capable of handling decimals backed by `fixed_len_byte_array` in which case the precision is limited by the size of the array. Since this is just a small wrapper around the underlying cudf code we can't change it right now.

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (razajafri) [reply java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:67]: Yes, its not the best. I thought about auto-generating this list when the parquet write method detects that it has decimals in the table. What do you think about that approach? 

I will think about it some more and think about future proofing this API

File: cpp/include/cudf/io/parquet.hpp
Comment (hyperbolic2346) [reply cpp/include/cudf/io/parquet.hpp:855]: `uint8_t` was selected in cudf because it seemed like there would never be a need for higher precision. I'm happy to adjust that if the need is there. I didn't consider the case of a `fixed_len_byte_array` since it isn't exactly supported in cudf currently.

The decimal precision array is a flattened array of values, meaning that if you have 4 decimal columns and 1 struct column with 2 decimals inside the array will need 6 values total. Each time we write out a column metadata for a decimal, we pop a value off the precision array and use that. The code will throw a logic error if too many or too few entries are provided.

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (revans2) [reply java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:67]: I really want an API that exposes some kind of a ParquetSchema so we can give the parquet writer the extra metadata about the types that it is missing. I am fine if we don't do it now. It sounds like some of that work is going to happen at the C++ layer so for now we probably can just keep this API so long as we document it properly for what "flattened" really means.

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (razajafri) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:3967]: I wanted to leave a few tests that tested the DEFAULT options. I can change them to use the Decimal columns if you want

File: cpp/include/cudf/io/parquet.hpp
Comment (harrism) [reply cpp/include/cudf/io/parquet.hpp:846]: This is the only file cpp-codeowners need to approve. I will approve once the copyright is changed to `2020-2021`

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (razajafri) [reply java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:75]: I have addressed the documentation on the method javadocs. 

Do you want me to create a follow-on issue for the Parquet-like API that you wanted?

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Review by karthikeyann — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:105]: The docs here need an explanation of flattened, either directly here or pointing to the example in `withPrecisionValues`.

File: java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetWriterOptions.java:61]: It would be good to mention that it's an in-order traversal of _all_ decimal columns, whether they are top-level columns in the table or nested within another column type like list or struct columns.  Bonus points if the example shows two decimal columns within the struct column, showing the relative ordering when that nested type is flattened.

**Review by jlowe — CHANGES_REQUESTED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by revans2** [comment]
@gpucibot merge

**Pull request merged**
